KYRA: Fix EoBCoreEngine::_hpIncrPerLevel table lookup

vrza opened the pull request:
KYRA: Fixed level-up hit point gain in eob

Eye of the Beholder uses a table for looking up HP gains when
characters level up. Table is EoBCoreEngine::_hpIncrPerLevel,
has 3 subsections of 6 entries each:

```
+const uint8 EoBCoreEngine::_hpIncrPerLevel[] = {
  10,   #0   # fighter HP die sides
   4,   #1   # mage HP die sides
   8,   #2   # cleric HP die sides
   6,   #3   # thief HP die sides
  10,   #4   # paladin? HP die sides
  10,   #5   # ranger? HP die sides
   9,   #6   # fighter max level to roll HP
  10,   #7   # mage max level to roll HP
   9,   #8   # cleric max level to roll HP
  10,   #9   # thief max level to roll HP
   9,  #10   # paladin? max level to roll HP 
   9,  #11   # ranger? max level to roll HP 
   3,  #12   # fighter static HP
   1,  #13   # mage static HP
   2,  #14   # cleric static HP
   2,  #15   # thief static HP
   3,  #16   # paladin? static HP
   3   #17   # ranger? static HP
};
```

The table is correct (I compared it to the original game manual).

However, when looking up max level to roll dice for HP bonus, and static HP bonuses after that level (table offsets 6 and 12 respectively), scummvm code was incorrectly using the character subclass index i (e.g. for Fighter/Thief, i = 0 for Fighter, i = 1 for Thief), instead of the character class type (fighter/mage/cleric/thief/etc.)

athrxx commented:
Thanks! I can confirm from disasm that the change is correct.

vrza commented:
@athrxx thanks! While we're here might be worth checking the surrounding code for another detail, regarding the constitution bonus. Currently, the constitution bonus (`m`) is unconditionally added to hit points on each level:

```cpp
	int m = getClassAndConstHitpointsModifier(c->cClass, c->constitutionCur);

	int h = 0;

	for (int i = 0; i < 3; i++) {
		if (!(levelIndex & (1 << i)))
			continue;

		int d = getCharacterClassType(c->cClass, i);

		if (c->level[i] <= _hpIncrPerLevel[6 + d])
			h += rollDice(1, (d >= 0) ? _hpIncrPerLevel[d] : 0);
		else
			h += _hpIncrPerLevel[12 + d];

		h += m;
	}
```

However, the AD&D rules say it should only be added when the HP die is rolled. AD&D 2nd Edition Player's Handbook, page 15:

> The Constitution bonus ends when a character reaches 10th level (9th for warriors and priests)-neither the Constitution bonus nor Hit Dice are added to a character‘s hit points after he has passed this level

So, to conform to AD&D rules, code should look like

```cpp
	int m = getClassAndConstHitpointsModifier(c->cClass, c->constitutionCur);

	int h = 0;

	for (int i = 0; i < 3; i++) {
		if (!(levelIndex & (1 << i)))
			continue;

		int d = getCharacterClassType(c->cClass, i);

		if (c->level[i] <= _hpIncrPerLevel[6 + d]) {
			h += m;
			h += rollDice(1, (d >= 0) ? _hpIncrPerLevel[d] : 0);
		} else {
			h += _hpIncrPerLevel[12 + d];
		}

	}
```

But this inaccuracy might be the way it was coded in the original game. If you find some time, can you try to compare with the disassembly?

P.S. To make this code easier to read, magic numbers 6 and 12 (offsets in the table) could be replaced with constants, e.g.

```cpp
	const unsigned int8 kOffsetHPRollDieSides = 0;
	const unsigned int8 kOffsetHPRollMaxLevel = 6;
	const unsigned int8 kOffsetStaticHPBonus = 12;
```

vrza commented:
Additionally, there is a rounding when casting to int issue here:

```cpp
h /= _numLevelsPerClass[c->cClass];
```

As an example, take a Warrior/Thief going from level 10 to level 12 in both classes.

Warrior static bonus is 3, but since the character is dual-class, he only gets + (3 / 1), which, when cast to int, equals +1, on level 11 and also +1 on level 12. Thief static bonus is 2, so dual-class character will get +1 on level 11 and +1 and level 12. In total, warrior/thief will gain (3/2 + 3/2 + 2/2 + 2/2) = 4 HP from lvl 10 to lvl 12, but in reality he should gain (3 + 3 + 2 + 2) / 2 = 5 HP.

This is not so easy to solve, but the simplest solution to implement would be e.g. to round down on odd levels and round up on even levels.

I can make a PR with all or any of the changes but let's first check the original game logic, and if we want to deviate from it by making such "improvements".

If the original game logic is lacking we could add a flag to ScummVM to use either the "original" the "improved" logic. But the changes  are far from game-breaking, and ScummVM already made some improvements (like adding multiple save slots to EotB 1), so it might be cool to make improvements without introducing a feature flag / optino.

athrxx commented:
Thanks for going over that code. 

I have checked what the original does:
- The constitution bonus `m` is always added (like in our code).
- It really only has that `h += rollDice(...)` in the `<=` code path
- The division you mentioned is like that, too.

So these discoveries really seem to qualify as improvements rather than actual bug fixes.
I wouldn't mind adding a feature flag for improved ADD logic. We have checkboxes like that for other engines, too.

I wouldn't want to change the default, since
- we ususally want to match the original and we do get bug reports about that sort of thing
- the code becomes kind of impossible to maintain/fix for me if I can't rely on the original code. It would basically mean that I have to work myself through the ADD handbook, too...

The rounding thing would really require some thought: How is the division supposed to be done in the ADD handbook? Are you supposed to keep the fraction and use it at the next level gain?

vrza commented:
Feature flag sounds good to me.

AD&D Handbook is not super clear on this, here's the quote:

> Later the character is likely to gain levels in different classes at different times. When this happens, roll the appropriate Hit Die and divide the result by the number of classes the character has (round fractions down, but a Hit Die never yields less than 1 hit point). The character's Constitution bonus is split between his classes, thus a fighter/mage gets 1/2 of his Con bonus when he goes up a level as a fighter and  the other 1/2 of the Con bonus when he goes up a level as a mage. A fighter/mage/thief would would get 1/3 of his bonus when he goes up as a    fighter, 1/3 when he goes up as a mage, and the other 1/3 when he goes up as a thief.

But I googled a bit and found someone mentioning this on a forum:

> The 2nd Edition product Demihuman Deities had a recommended method that prevents the loss of fractional hit points. It recommends keeping a running total of your rolls (i.e. level 1: 6, 4; level 2: 3, 2; level 3: 4), adding the appropriate Constitution bonus to each die (keeping in mind that a high-Con Fighter could get +4 to the Fighter class, but only +2 to his other class), totaling it all up, and dividing by the number of classes. Every time you level up, you add the new roll to the back-end total, and re-divide. In this way, the character will always have    exactly half or one third of the total of his rolls, and will never have a steady loss from rounding.

While this should be verified from the reference ("Demihuman Deities" book), it also complicates the save game format, as more state would need to be saved (HP die roll at every previous level). That's why I suggested rounding up and down as an easy to implement compromise, as it doesn't require us to track any additional state.

athrxx commented:
I have added the checkbox and also put in your suggestion from above (so that I actually have something to do for the checkbox).
Please check if that goes into the right direction for you.

I haven't done anything about the lost level division fraction yet.

And please feel free to make another PR if you want to. I just didn't want to dump the checkbox task on you, since I can do that rather quickly while you would probably need some time to understand how that works in ScummVM.

vrza commented:
@athrxx Good stuff, thanks for adding the checkbox and flag, we are definitely on the same page.

Please see PR #5957. I moved the enhanced hit point gain rules to a separate function, to make it easier to read and maintain the code, and to promote good code factoring. Also added fix for fractions by alternate rounding up and down, with a TODO to improve this in the future.

Does it make sense to you to add this checkbox to the in-game menu as well?
